src/components/include/sl.h

```diff
@@ -382,13 +404,24 @@ sl_cs_exit_schedule_nospin_arg(struct sl_thd *to)
 		if (unlikely(!pt)) t = sl__globals()->idle_thd;
 		else               t = sl_mod_thd_get(pt);
 	}
-	thdcap = t->thdcap;
-	prio   = t->prio;
 
+	if (t->flags & SL_FLAG_OWNTC) {
+		assert(t->budget && t->period);
+
+		if (t->last_replenish == 0 || t->last_replenish + t->period <= now) {
+			tcap_res_t currbudget;
+
+			t->last_replenish = now;
+			currbudget        = (tcap_res_t)cos_introspect(ci, sl_thd_tcap(t), TCAP_GET_BUDGET);
+			/* TODO: need to change logic for SNDCAP with tcap_delegate, and error handling */
+			if (currbudget < t->budget && cos_tcap_transfer(sl_thd_rcvcap(t), sl__globals()->sched_tcap, (t->budget - currbudget), t->prio)) assert(0);
```

Comment:
My bad, it should not be an assertion. Especially if we try to transfer from within child scheduler having finite budget.